service/debugger/debugger.go

```diff
@@ -1410,10 +1464,91 @@ func traverse(t proc.ValidTargets, f *proc.Function, depth int, followCalls int)
 		}
 		f := parent.Func
 		text, err := proc.Disassemble(t.Memory(), nil, t.Breakpoints(), t.BinInfo(), f.Entry, f.End)
+
 		if err != nil {
 			return nil, fmt.Errorf("disassemble failed with error %w", err)
 		}
 		for _, instr := range text {
+			if f.Name == "runtime.deferreturn" {
+				// Defer functions are called in a special way wherein the destination location is nil
+				// Hence its required to put a breakpoint inorder to acquire the address of the function
+				// at runtime and we do this via a call back mechanism
+				if instr.IsCall() && instr.DestLoc == nil {
+					deferbp, err := t.SetBreakpoint(0, instr.Loc.PC, proc.NextBreakpoint, nil)
+					deferbp.RootFuncName = rootstr
+					if err != nil {
+						fmt.Printf("error setting breakpoint inside deferreturn\n")
+					}
+					deferbrklet := deferbp.Breaklets[len(deferbp.Breaklets)-1]
+					deferbrklet.Callback = func(th proc.Thread, _ *proc.Target) (bool, error) {
+						rawlocs, err := proc.ThreadStacktrace(t.Group.Selected, t.CurrentThread(), 20)
+						if err != nil {
+							fmt.Printf("Thread stack trace returned error\n")
+						}
+						bpi := &api.BreakpointInfo{}
+						// Since the defer function is known only at runtime, the depth is likewise
+						// calculated by referring to the stack and the mechanism is exactly the same
+						// as that used in pkg/terminal/command.go:printTraceOutput
+						bpi.Stacktrace, err = d.convertStacktrace(rawlocs, nil)
+						if err != nil {
+							fmt.Printf("convert stack trace returned error\n")
+						}
+						stack := bpi.Stacktrace
+						rootindex := -1
+						for i := len(stack) - 1; i >= 0; i-- {
+							if stack[i].Function.Name() == deferbp.RootFuncName {
+								if rootindex == -1 {
+									rootindex = i
+									break
+								}
+							}
+						}
+						sdepth := rootindex + 1
+
+						if sdepth+1 > followCalls {
+							return false, nil
+						}
+						regs, err := th.Registers()
+						if err != nil {
+							fmt.Printf("registers inside callback returned err\n")
+
+						}
+						dregs := t.Group.Selected.BinInfo().Arch.RegistersToDwarfRegisters(0, regs)
+						dregs.Reg(^uint64(0))
+						for i := 0; i < dregs.CurrentSize(); i++ {
+							reg := dregs.Reg(uint64(i))
+							if reg == nil {
+								continue
+							}
+							name, _, value := t.Group.Selected.BinInfo().Arch.DwarfRegisterToString(i, reg)
+							if isDeferCallReg(t.Group.Selected, name) {
+								addr, err := strconv.ParseUint(value, 0, 64)
+								if err != nil {
+									fmt.Printf("error parsing function address\n")
+								}
+								fn := t.Group.Selected.BinInfo().PCToFunc(addr)
+								_, err = createFnTracepoint(d, fn.Name, rootstr, followCalls)
```

Comment:
You shouldn't create additional breakpoints from within a breaklet callback, it will modify the breaklet array as it is being scanned. I don't know if it actually leads to problem in this particular case, but it's too brittle.